core/src/main/java/com/datastrato/gravitino/lock/LockManager.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2023 Datastrato Pvt Ltd.
+ * This software is licensed under the Apache License version 2.
+ */
+
+package com.datastrato.gravitino.lock;
+
+import com.datastrato.gravitino.NameIdentifier;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.util.Stack;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * LockManager is a lock manager that manages the lock of the resource path. It will lock the whole
+ * path from root to the resource path.
+ *
+ * <p>Assuming we need to alter the table `metalake.catalog.db1.table1`, the lock manager will lock
+ * the following
+ *
+ * <pre>
+ *   /                                    readLock
+ *   /metalake                            readLock
+ *   /metalake/catalog                    readLock
+ *   /metalake/catalog/db1                readLock
+ *   /metalake/catalog/db/table1          writeLock
+ * </pre>
+ *
+ * If the lock manager fails to lock the resource path, it will release all the locks that have been
+ * locked.
+ */
+public class LockManager {
+  private static final Logger LOG = LoggerFactory.getLogger(LockManager.class);
+  private static final long TIME_AFTER_LAST_ACCESS_TO_EVICT_IN_MS = 24 * 60 * 60 * 1000; // 1 day
+
+  @VisibleForTesting final TreeLockNode treeLockRootNode = new TreeLockNode("root");
+  private final ThreadLocal<Stack<TreeLockPair>> holdingLocks = ThreadLocal.withInitial(Stack::new);
```

Comment:
What is the purpose of using thread local stack?

Reply:
It's to keep the locks this thread holds and then we can use it to unlock them.

Reply:
I think you don't have to maintain a stack if you also store the "parent" field in the node.

Reply:
I have reviewed the code here. I use a thread-local stack to store locks for the following reasons:
- It's easy to manage. Each thread has a thread-local variant to keep the locks it holds and then unlock them.
- To avoid potential exceptions during the lock procedure, as the tree lock is crucial. We need to ensure that every corner is in good condition. For example, what if the operation on the map throws an exception for the following code? We may need to determine which node has been locked and which hasn't.

```java
 try {
      for (String level : levels) {
        lockNode = lockNode.getOrCreateChild(level); // could throw exception
        lockNode.lock(LockType.READ);
        holdingLocks.get().push(TreeLockPair.of(lockNode, LockType.READ));
      }

      lockNode = lockNode.getOrCreateChild(identifier.name());
      lockNode.lock(lockType);
      holdingLocks.get().push(TreeLockPair.of(lockNode, lockType));
    } catch (Exception e) {
      unlockResourcePath();
      LOG.error("Failed to lock resource path {}", identifier, e);
      throw e;
    }
``` 

- Adding a field `parent` solely is NOT very convenient to release locks, we need to traverse to the leaf node from the root by name and then unlock them inversely. This doesn't seem very efficient.

@jerryshao 
Do you have any other considerations about using the `parent` field here? Maybe I missed something.